lib/segment/src/segment/facet.rs

```diff
@@ -58,15 +75,18 @@ impl Segment {
 
                 Either::Left(iter)
             } else {
-                // go over the values and filter the points
-                // aka. read from facet index
-                //
-                // This is more similar to a full-scan, but we won't be hashing so many times.
-                context = payload_index.struct_filtered_context(filter);
+                // Use cardinality estimations for each value to estimate the counts
+                let cardinality_ratio =
+                    filter_cardinality.exp as f32 / self.available_point_count() as f32;
 
+                // Go over each value in the map index
                 let iter = facet_index
-                    .iter_filtered_counts_per_value(&context)
+                    .iter_counts_per_value()
                     .check_stop(|| is_stopped.load(Ordering::Relaxed))
+                    .map(move |hit| FacetHit {
+                        value: hit.value,
+                        count: (hit.count as f32 * cardinality_ratio).round() as usize,
```

Comment:
Do I understand this correct, that this approach assumes that filter doesn't correlate with facet value?

Reply:
Hmm good point. It does not take into consideration that the facet key may be included in the filter

Reply:
Not only that. Facet values can be biased towards filtering. I don't think we can make independence assumption in this case.

I would rather make sampling like this:

- sample random points -> filter them -> compute facets -> repeat until sample size is either big enough, or reach max attempts

Reply:
And then adjust the sampled counts with the cardinality estimation?